server/chat-plugins/friends.ts

```diff
@@ -0,0 +1,915 @@
+/**
+ * Friends list plugin.
+ * Allows for adding and removing friends, as well as seeing their activity.
+ * Written by Mia.
+ * @author mia-pi-git
+ */
+import type {Database, Transaction, Statement} from 'better-sqlite3';
+import {Utils} from '../../lib/utils';
+import {FS} from '../../lib/fs';
+import {QueryProcessManager} from '../../lib/process-manager';
+import {Repl} from '../../lib/repl';
+import {Config} from '../config-loader';
+
+/** Max friends per user */
+const MAX_FRIENDS = 100;
+/** Max friend requests. */
+const MAX_REQUESTS = 6;
+const REQUEST_EXPIRY_TIME = 30 * 24 * 60 * 60 * 1000;
+const MAX_PROCESSES = 1;
+
+interface DatabaseRequest {
+	statement: string;
+	type: 'all' | 'get' | 'run' | 'transaction';
+	data: AnyObject | any[];
+}
+
+interface DatabaseResult {
+	/** if there's an error, this is it. */
+	error?: string;
+	result?: any;
+}
+
+const STATUS_COLORS: {[k: string]: string} = {
+	idle: '#ff7000',
+	online: '#009900',
+	busy: '#cc3838',
+};
+
+const STATUS_TITLES: {[k: string]: string} = {
+	online: 'Online',
+	idle: 'Idle',
+	busy: 'Busy',
+	offline: 'Offline',
+};
+
+export const Friends = new class {
+	/** `Map<oldID, newID> */
+	renames: Map<string, string>;
+	database?: Database;
+	/** `Map<transferTo, transferFrom>` */
+	transferRequests: Map<string, string>;
+	constructor() {
+		this.setupDatabase();
+		this.renames = this.getRenames();
+		const existingRequests = 'Chat' in global ? Chat.oldPlugins.friends?.transferRequests : undefined;
+		this.transferRequests = existingRequests || new Map();
+		void this.checkExpiringRequests();
+	}
+	setupDatabase() {
+		if (!Config.usesqlite) return;
+		const Sqlite = require('better-sqlite3');
+		try {
+			this.database = new Sqlite(`${__dirname}/../../databases/friends.db`, {fileMustExist: true});
+		} catch (e) {
+			this.database = new Sqlite(`${__dirname}/../../databases/friends.db`);
+			this.database!.exec(FS('databases/schemas/friends.sql').readSync());
+		}
+	}
+	async checkExpiringRequests() {
+		if (!this.database) return;
+		// requests expire after one month. this is checked both on intialization
+		// (hotpatch, to ensure accuracy in case they both don't log in)
+		// and when the user the request is sent to / sent from logs in.
+		const results = this.database.prepare(`SELECT * FROM friend_requests`).all();
+		const removed = [];
+		for (const request of results) {
+			if ((Date.now() - request.sent_at) > REQUEST_EXPIRY_TIME) {
+				// expires after a month
+				await this.removeRequest(request.receiver, request.sender);
+				removed.push(request);
+			}
+		}
```

Comment:
Correlated queries must be in transaction, and you're making `O(n)` queries here. Just do a `DELETE from friend_requests WHERE <expiry condition>`